### PR TITLE
[MSFT] Introduce msft.module.extern op for PyCDE lowering.

### PR DESCRIPTION
**Pull request by mikeurbach**
The op is identical to hw.module.extern, but it can be referred to by
a msft.instance op. This adds a cleaner layering where PyCDE only
generates msft.module/msft.module.extern/msft.instance ops. Before, it
would generate msft.module/hw.module.extern, leading to a mix of
msft.instance/hw.instance for LowerMSFTToHW to consume. This change
introduces a clear contract for PyCDE's output: all module and
instance ops are from the MSFT dialect.

Some of the implementation is duplicated into MSFTOps.cpp from
HWOps.cpp, but refactoring is left as a future exercise, related to
https://llvm.discourse.group/t/instancelike-and-modulelike-interfaces/4132/3.

### CODE REVIEW

**Comment by mikeurbach** [comment]
If this PR has grown too large, I can split it up into a few smaller PRs. I wanted to get it working end-to-end before I sent out the earlier pieces for review.

File: frontends/PyCDE/src/pycde/module.py [hunk 1]
```diff
@@ -204,9 +204,14 @@ def instantiate(self, instance_name: str, inputs: dict, loc):
     if self.extern_name is None:
       return self.circt_mod.create(instance_name, **inputs, loc=loc)
     else:
+      paramdecl_list = [
+          hw.ParamDeclAttr.get(p.name, mlir.ir.TypeAttr.get(p.attr.type),
```
Comment (teqdruid) [inline frontends/PyCDE/src/pycde/module.py:208]: Where were we doing this previously?

**Review by teqdruid — APPROVED** [review]
This'll also come in handy for adding a physical location attribute.

It's rather unfortunate that the parsing/printing stuff isn't generic. When I created msft.module/instance I extracted out and made accessible some of that parsing code, but it took awhile. We should discuss how to share code like this at some point rather than copy-paste-modify.

**Comment by teqdruid** [comment]
#2107

File: frontends/PyCDE/src/pycde/module.py (hunk 1 above)
Comment (mikeurbach) [reply frontends/PyCDE/src/pycde/module.py:208]: It was in the _hw_ops_ext before... I will move this bit to _msft_ops_ext so we don't have to worry about it everywhere.

**Pull request merged**